<a href="https://colab.research.google.com/github/jagangirisaballa/iNeuron-Assignments/blob/master/Jagan_Data_Cleaning_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import pandas as pd
import numpy as np


It happens all the time: someone gives you data containing malformed strings, Python, lists and missing data. How do you tidy it up so you can get on with the analysis? Take this monstrosity as the DataFrame to use in the following puzzles: 


In [0]:
df = pd.DataFrame({'From_To': ['LoNDon_paris', 'MAdrid_miLAN', 'londON_StockhOlm', 
'Budapest_PaRis', 'Brussels_londOn'], 
'FlightNumber': [10045, np.nan, 10065, np.nan, 10085], 
'RecentDelays': [[23, 47], [], [24, 43, 87], [13], [67, 32]], 
'Airline': ['KLM(!)', '<Air France> (12)', '(British Airways. )', 
'12. Air France', '"Swiss Air"']}) 


Checking the Nature of Data

In [26]:
df.head(15)

From_To  FlightNumber  RecentDelays              Airline
0      LoNDon_paris       10045.0      [23, 47]               KLM(!)
1      MAdrid_miLAN           NaN            []    <Air France> (12)
2  londON_StockhOlm       10065.0  [24, 43, 87]  (British Airways. )
3    Budapest_PaRis           NaN          [13]       12. Air France
4   Brussels_londOn       10085.0      [67, 32]          "Swiss Air"

Checking the Type of Data

In [27]:
df.describe()

FlightNumber
count           3.0
mean        10065.0
std            20.0
min         10045.0
25%         10055.0
50%         10065.0
75%         10075.0
max         10085.0

In [28]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   From_To       5 non-null      object 
 1   FlightNumber  3 non-null      float64
 2   RecentDelays  5 non-null      object 
 3   Airline       5 non-null      object 
dtypes: float64(1), object(3)
memory usage: 288.0+ bytes


Some values in the the FlightNumber column are missing. These numbers are meant to increase by 10 with each row so 10055 and 10075 need to be put in place. Fill in these missing numbers and make the column an integer column (instead of a float column). 


In [29]:
df['FlightNumber'] = df['FlightNumber'].interpolate().astype(int)
df

From_To  FlightNumber  RecentDelays              Airline
0      LoNDon_paris         10045      [23, 47]               KLM(!)
1      MAdrid_miLAN         10055            []    <Air France> (12)
2  londON_StockhOlm         10065  [24, 43, 87]  (British Airways. )
3    Budapest_PaRis         10075          [13]       12. Air France
4   Brussels_londOn         10085      [67, 32]          "Swiss Air"

The From_To column would be better as two separate columns! Split each string on the underscore delimiter _ to give a new temporary DataFrame with the correct values. Assign the correct column names to this temporary DataFrame

In [30]:
temp = df.From_To.str.split('_', expand=True)
temp.columns = ['From', 'To']
temp

From         To
0    LoNDon      paris
1    MAdrid      miLAN
2    londON  StockhOlm
3  Budapest      PaRis
4  Brussels     londOn

Notice how the capitalisation of the city names is all mixed up in this temporary DataFrame. Standardise the strings so that only the first letter is uppercase (e.g. "londON" should become "London".) 


In [31]:
temp['From'] = temp['From'].str.capitalize()
temp['To'] = temp['To'].str.capitalize()
temp

From         To
0    London      Paris
1    Madrid      Milan
2    London  Stockholm
3  Budapest      Paris
4  Brussels     London

Delete the From_To column from df and attach the temporary DataFrame from the previous questions. 


In [32]:
df = df.drop('From_To', axis=1)
df = df.join(temp)
df

FlightNumber  RecentDelays              Airline      From         To
0         10045      [23, 47]               KLM(!)    London      Paris
1         10055            []    <Air France> (12)    Madrid      Milan
2         10065  [24, 43, 87]  (British Airways. )    London  Stockholm
3         10075          [13]       12. Air France  Budapest      Paris
4         10085      [67, 32]          "Swiss Air"  Brussels     London

In the RecentDelays column, the values have been entered into the DataFrame as a list. We would like each first value in its own column, each 
second value in its own column, and so on. If there isn't an Nth value, the value should be NaN. 


In [33]:
delays = df['RecentDelays'].apply(pd.Series)
delays

0     1     2
0  23.0  47.0   NaN
1   NaN   NaN   NaN
2  24.0  43.0  87.0
3  13.0   NaN   NaN
4  67.0  32.0   NaN

In [35]:
delays.columns = ['delay_{}'.format(n) for n in range(1, len(delays.columns)+1)]
df = df.drop('RecentDelays', axis=1).join(delays)
df

FlightNumber          Airline      From  ... delay_1  delay_2  delay_3
0         10045              KLM    London  ...    23.0     47.0      NaN
1         10055       Air France    Madrid  ...     NaN      NaN      NaN
2         10065  British Airways    London  ...    24.0     43.0     87.0
3         10075       Air France  Budapest  ...    13.0      NaN      NaN
4         10085        Swiss Air  Brussels  ...    67.0     32.0      NaN

[5 rows x 7 columns]

In [36]:
df['Airline'] = df['Airline'].str.extract('([a-zA-Z\s]+)', expand=False).str.strip()
# note: using .strip() gets rid of any leading/trailing spaces
df

FlightNumber          Airline      From  ... delay_1  delay_2  delay_3
0         10045              KLM    London  ...    23.0     47.0      NaN
1         10055       Air France    Madrid  ...     NaN      NaN      NaN
2         10065  British Airways    London  ...    24.0     43.0     87.0
3         10075       Air France  Budapest  ...    13.0      NaN      NaN
4         10085        Swiss Air  Brussels  ...    67.0     32.0      NaN

[5 rows x 7 columns]